crates/noirc_evaluator/src/brillig/brillig_gen.rs

```diff
@@ -37,11 +40,23 @@ impl BrilligGen {
             return *register_index;
         }
 
-        let register = RegisterIndex::from(self.latest_register);
+        let register = self.create_register();
+
+        // Cache the `ValueId` so that if we call it again, it will
+        // return the register that has just been created.
+        //
+        // WARNING: This assumes that a register has not been
```

Comment:
This is pretty sneaky, especially with function calls which may overwrite registers using MOV. MOV instruction does not update the cache

Reply:
With no recursion this should be okay, the edge case to consider is function calls which may overwrite registers for the function that called it. I opened up a milestone for function to function call, so this should be exposed there